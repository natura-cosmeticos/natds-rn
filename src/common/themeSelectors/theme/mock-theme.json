{
  "avatarSizes": {
    "tiny": {
      "size": 24,
      "fontSize": ".625rem"
    },
    "small": {
      "size": 32,
      "fontSize": ".875rem"
    },
    "standard": {
      "size": 40,
      "fontSize": "1rem"
    },
    "large": {
      "size": 48,
      "fontSize": "1rem"
    },
    "huge": {
      "size": 80,
      "fontSize": "1.5rem"
    }
  },
  "buttonSizes": {
    "small": {
      "paddingTop": 8,
      "paddingRight": 8,
      "paddingBottom": 8,
      "paddingLeft": 8,
      "minHeight": 32
    },
    "medium": {
      "paddingTop": 12,
      "paddingRight": 16,
      "paddingBottom": 12,
      "paddingLeft": 16,
      "minHeight": 40
    },
    "large": {
      "paddingTop": 16,
      "paddingRight": 16,
      "paddingBottom": 16,
      "paddingLeft": 16,
      "minHeight": 48
    }
  },
  "fontSize": {
    "h1": {
      "fontSize": "6rem",
      "fontWeight": 300
    },
    "h2": {
      "fontSize": "3.75rem",
      "fontWeight": 300
    },
    "h3": {
      "fontSize": "3rem",
      "fontWeight": 400
    },
    "h4": {
      "fontSize": "2.125rem",
      "fontWeight": 400
    },
    "h5": {
      "fontSize": "1.5rem",
      "fontWeight": 400
    },
    "h6": {
      "fontSize": "1.25rem",
      "fontWeight": 600
    },
    "subtitle1": {
      "fontSize": "1rem",
      "fontWeight": 400
    },
    "subtitle2": {
      "fontSize": "0.875rem",
      "fontWeight": 500
    },
    "body1": {
      "fontSize": "1rem",
      "fontWeight": 400
    },
    "body2": {
      "fontSize": "0.875rem",
      "fontWeight": 400
    },
    "button": {
      "fontSize": "0.875rem",
      "fontWeight": 600
    },
    "caption": {
      "fontSize": "0.75rem",
      "fontWeight": 400
    },
    "overline": {
      "fontSize": "0.75rem",
      "fontWeight": 400
    }
  },
  "iconSizes": {
    "micro": 16,
    "tiny": 24,
    "small": 32,
    "standard": 48,
    "medium": 64,
    "large": 72,
    "largex": 96,
    "largexx": 128,
    "huge": 144,
    "hugex": 192,
    "hugexx": 256
  },
  "opacity": {
    "12": 0.12,
    "16": 0.16,
    "24": 0.24,
    "32": 0.32,
    "48": 0.48,
    "56": 0.56,
    "64": 0.64,
    "80": 0.8,
    "00": 0,
    "04": 0.04,
    "08": 0.08
  },
  "overlay": {
    "solidBlack": 0.56
  },
  "radius": {
    "none": 0,
    "circle": "50%",
    "small": 2,
    "medium": 4,
    "large": 8
  },
  "sizes": {
    "none": 0,
    "micro": 4,
    "tiny": 8,
    "small": 16,
    "standard": 24,
    "semi": 32,
    "semix": 40,
    "medium": 48,
    "mediumx": 56,
    "large": 64,
    "largex": 72,
    "largexx": 80,
    "largexxx": 88,
    "huge": 96,
    "hugex": 128,
    "hugexx": 144,
    "hugexxx": 192,
    "veryhuge": 256
  },
  "spacing": {
    "spacing": 8,
    "spacingNone": 0,
    "spacingMicro": 4,
    "spacingTiny": 8,
    "spacingSmall": 16,
    "spacingStandard": 24,
    "spacingSemi": 32,
    "spacingLarge": 48,
    "spacingHuge": 64
  },
  "colorTokens": {
    "colorPrimary": "#F4AB34",
    "colorOnPrimary": "#333333",
    "colorPrimaryLight": "#FEFDE8",
    "colorOnPrimaryLight": "#333333",
    "colorPrimaryDark": "#EF8426",
    "colorOnPrimaryDark": "#333333",
    "colorSecondary": "#FF6B0B",
    "colorOnSecondary": "#000000",
    "colorSecondaryLight": "#FFF8E1",
    "colorOnSecondaryLight": "#000000",
    "colorSecondaryDark": "#FF5808",
    "colorOnSecondaryDark": "#000000",
    "colorBackground": "#FAFAFA",
    "colorOnBackground": "#333333",
    "colorSurface": "#FFFFFF",
    "colorOnSurface": "#333333",
    "colorHighlight": "#000000",
    "colorHighEmphasis": "#333333",
    "colorMediumEmphasis": "#777777",
    "colorLowEmphasis": "#BBBBBB",
    "colorLink": "#227BBD",
    "colorOnLink": "#FFFFFF",
    "colorSuccess": "#569A32",
    "colorOnSuccess": "#FFFFFF",
    "colorWarning": "#FCC433",
    "colorOnWarning": "#333333",
    "colorAlert": "#E74627",
    "colorOnAlert": "#FFFFFF"
  },
  "elevation": {
    "1": {
      "elevation": 1,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 1, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 1
    },
    "2": {
      "elevation": 2,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 2, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 2
    },
    "3": {
      "elevation": 3,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 3, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 4
    },
    "4": {
      "elevation": 4,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 4, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 5
    },
    "5": {
      "elevation": 5,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 6, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 10
    },
    "6": {
      "elevation": 6,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 8, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 10
    },
    "7": {
      "elevation": 7,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 9, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 12
    },
    "8": {
      "elevation": 8,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 12, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 17
    },
    "9": {
      "elevation": 9,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 16, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 24
    },
    "10": {
      "elevation": 10,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 24, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 38
    }
  }
}
