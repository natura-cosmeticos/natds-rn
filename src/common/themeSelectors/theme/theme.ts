import { tokens as styleTokens } from '@naturacosmeticos/natds-styles';
import { IColorThemeTokens } from '@naturacosmeticos/natds-styles/dist/tokens/colors/themeBound/IColorThemeTokens';
import { IElevation } from '@naturacosmeticos/natds-styles/dist/tokens/elevation/IElevation';

type Tokens = typeof styleTokens;
type TokensWithoutColor = Omit<Tokens, 'colors'>
type TokensWithoutColorTokens = Omit<TokensWithoutColor, 'colorTokens'>
type TokensWithoutElevation = Omit<TokensWithoutColorTokens, 'elevation'>

interface ElevationProps {
  elevation: number
  shadowColor: string
  shadowOffset: { height: number, width: number }
  shadowOpacity: number
  shadowRadius: number
}

type Elevation = Record<keyof IElevation, ElevationProps>


export type Theme = TokensWithoutElevation
  & {
    colorTokens: IColorThemeTokens,
    elevation: Elevation | {}
  }

type ThemeMode = 'dark' | 'light';

const parseElevation = (elevation: string) => {
  const regEx = /(\d+(\.\d+)?)/g;
  const matches = elevation.match(regEx);

  if (matches) {
    const [width, height, blurRadius, , r, g, b, alpha] = matches;

    const toUnit = (prop: string) => parseInt(prop, 0);

    return {
      elevation: 2,
      shadowColor: `rgb(${r}, ${g}, ${b})`,
      shadowOffset: { height: toUnit(height), width: toUnit(width) },
      shadowOpacity: parseFloat(alpha),
      shadowRadius: toUnit(blurRadius),
    };
  }

  return {};
};

const buidElevation = (elevation: IElevation) => {
  const nativeElevation = {};

  Object
    .keys(elevation)
    .forEach((key) => { nativeElevation[key] = parseElevation(elevation[key]); });

  return nativeElevation;
};

export const buildTheme = (tokens: Tokens, brand: string, mode: ThemeMode) => {
  const colorThemeName = `${brand}${mode === 'light' ? 'Light' : 'Dark'}ColorTokens`;
  const localTokens = { ...tokens };

  delete localTokens.colors;

  const theme: Theme = {
    ...localTokens,
    colorTokens: localTokens.colorTokens[colorThemeName],
    elevation: buidElevation(localTokens.elevation),
  };

  return theme;
};

interface ThemeSelector {
  (theme: Theme): string | undefined
}

export const getTheme = (selector: ThemeSelector) => (
  { theme }: { theme: Theme },
) => selector(theme);
