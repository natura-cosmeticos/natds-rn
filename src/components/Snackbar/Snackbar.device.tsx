import { storiesOf } from '@storybook/react-native'
import {
  Colors,
  Interactive
} from './Snackbar.stories'

storiesOf('Snackbar', module)
  .add('Colors', Colors)
  .add('Interactive', Interactive)
