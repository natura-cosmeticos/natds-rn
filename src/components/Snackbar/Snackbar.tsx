import React, { useEffect, useRef, useState } from 'react';
import { Animated, Platform } from 'react-native';
import { withTheme } from 'styled-components/native';
import { Theme, getShadowBySize } from '../../common/themeSelectors';
import {
  SnackbarWrapper, SnackbarText, SnackbarButtonWrapper, getColorByType,
} from './Snackbar.styles';
import { ButtonBase } from '../Button/ButtonBase';

export type SnackbarType = 'standard' | 'success' | 'error' | 'warning' | 'info';

export interface SnackbarProps {
  /**
   * Theme
   */
  theme: Theme;
  /**
   * Snackbar's message content
   */
  message: string;
  /**
   * Snackbar's button text content when available
   */
  buttonText?: string;
  /**
   * Snackbar's type
   */
  type?: SnackbarType;
  /**
   * Snackbar's shown duration.
   */
  autoHideDuration?: number;
  /**
   * Whether the Snackbar is open
   */
  open: boolean;
  /**
   * How many message lines to be shown before clipping
   */
  numberOfLines?: number;
  /**
   * Callback when Snackbar is closed. The 'open' prop needs to be updated when this is called
   */
  onClose?: () => void;
}

// eslint-disable-next-line max-statements
export const SnackbarComponent = ({
  theme,
  message,
  buttonText,
  type = 'standard',
  autoHideDuration = 5000,
  open,
  numberOfLines = 2,
  onClose,
}: SnackbarProps) => {
  const fadeAnim = useRef(new Animated.Value(0)).current;
  const [isOpen, setIsOpen] = useState(false);
  const [isMultiLineText, setIsMultiLineText] = useState(false);
  const autoHideTimer = React.useRef<NodeJS.Timeout>();

  const handleClose = () => {
    if (onClose) onClose();
  };

  const show = () => {
    setIsOpen(true);
    // fade in duration recommended by material-ui
    Animated.timing(fadeAnim, { duration: 225, toValue: 1, useNativeDriver: Platform.OS !== 'web' })
      .start(({ finished }) => {
        if (finished) autoHideTimer.current = global.setTimeout(handleClose, autoHideDuration);
      });
  };

  const hide = () => {
    // fade out duration recommended by material-ui
    Animated.timing(fadeAnim, { duration: 195, toValue: 0, useNativeDriver: Platform.OS !== 'web' })
      .start(({ finished }) => {
        if (finished) setIsOpen(false);
      });
  };

  const toggle = () => {
    if (open) return show();

    return hide();
  };

  useEffect(() => {
    toggle();

    return () => {
      if (autoHideTimer.current) global.clearTimeout(autoHideTimer.current);
    };
  }, [open, autoHideDuration]);

  if (!isOpen) return null;

  return (
    <SnackbarWrapper
      type={type}
      as={Animated.View}
      style={{ opacity: fadeAnim, ...getShadowBySize(theme, 'large') }}
      testID="natds-snackbar-wrapper"
    >
      <SnackbarText
        type={type}
        testID="natds-snackbar-text"
        numberOfLines={numberOfLines}
        onTextLayout={({ nativeEvent: { lines } }) => {
          setIsMultiLineText(lines.length >= numberOfLines);
        }}
        ellipsizeMode="tail"
        isTwoLineAction={isMultiLineText && !!buttonText}
      >
        {message}
      </SnackbarText>
      {buttonText && (
        <SnackbarButtonWrapper isTwoLineAction={isMultiLineText && !!buttonText}>
          <ButtonBase
            text={buttonText}
            type="text"
            onPress={handleClose}
            testID="natds-snackbar-button"
            size="medium"
            textLabelStyle={{ color: getColorByType(theme, type) }}
          />
        </SnackbarButtonWrapper>
      )}
    </SnackbarWrapper>
  );
};

export const Snackbar = withTheme(SnackbarComponent);
