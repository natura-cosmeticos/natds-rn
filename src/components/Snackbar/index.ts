export { Snackbar, SnackbarProps } from './Snackbar'
