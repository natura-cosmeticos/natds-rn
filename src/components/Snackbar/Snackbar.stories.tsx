import React, { useState } from 'react'
import { View, Button } from 'react-native'
import { text, select, number } from '@storybook/addon-knobs'
import { Snackbar, SnackbarType } from './Snackbar'
import { StoryContainer, StoryWrapper } from '../../common/HelperComponents/StoryContainer'

const description = () => `
- - -

  ### NOTE for UX: This component is available in the following variants:

  - ✅ Standard

With the following attribute status:

  - **Color:**
    - ✅ Error
    - ✅ Info
    - ✅ Standard
    - ✅ Success
    - ✅ Warning

- - -
`

export default {
  component: Snackbar,
  parameters: {
    componentSubtitle: 'Snackbars provide brief messages about app processes at the bottom of the screen.',
    docs: {
      extractComponentDescription: description
    }
  },
  title: 'Components|Snackbar'
}

const snackbarTypeOptions = ['error', 'info', 'standard', 'success', 'warning']

export const Colors = () => {
  const [standard, setStandard] = useState(false)
  const [info, setInfo] = useState(false)
  const [warning, setWarning] = useState(false)
  const [error, setError] = useState(false)
  const [success, setSuccess] = useState(false)
  const [multi, setMulti] = useState(false)
  const [withButton, setWithButton] = useState(false)

  const defaultMessage = 'A short snackbar message'
  const multiMessage = 'A longer message to hopefully show how it behaves when there is a multiline snackbar. The text will be truncated and the missing text at the end of the line will be indicated by an ellipsis glyph'

  return (
    <StoryWrapper title="Variants">
      <StoryContainer title="Standard">
        <View style={{ backgroundColor: 'gainsboro', flex: 1, height: 200 }}>
          <Snackbar message={defaultMessage} open={standard} onClose={() => setStandard(false)} />
          <Button onPress={() => setStandard(!standard)} title="open standard snackbar" />
        </View>
      </StoryContainer>

      <StoryContainer title="Info">
        <View style={{ backgroundColor: 'gainsboro', flex: 1, height: 200 }}>
          <Snackbar message={defaultMessage} open={info} onClose={() => setInfo(false)} type="info" />
          <Button onPress={() => setInfo(!info)} title="open info snackbar" />
        </View>
      </StoryContainer>

      <StoryContainer title="Warning">
        <View style={{ backgroundColor: 'gainsboro', flex: 1, height: 200 }}>
          <Snackbar message={defaultMessage} open={warning} onClose={() => setWarning(false)} type="warning" />
          <Button onPress={() => setWarning(!warning)} title="open warning snackbar" />
        </View>
      </StoryContainer>

      <StoryContainer title="Error">
        <View style={{ backgroundColor: 'gainsboro', flex: 1, height: 200 }}>
          <Snackbar message={defaultMessage} open={error} onClose={() => setError(false)} type="error" />
          <Button onPress={() => setError(!error)} title="open error snackbar" />
        </View>
      </StoryContainer>

      <StoryContainer title="Success">
        <View style={{ backgroundColor: 'gainsboro', flex: 1, height: 200 }}>
          <Snackbar message={defaultMessage} open={success} onClose={() => setSuccess(false)} type="success" />
          <Button onPress={() => setSuccess(!success)} title="open success snackbar" />
        </View>
      </StoryContainer>

      <StoryContainer title="Multiline">
        <View style={{ backgroundColor: 'gainsboro', flex: 1, height: 200 }}>
          <Snackbar message={multiMessage} open={multi} onClose={() => setMulti(false)} type="standard" />
          <Button onPress={() => setMulti(!multi)} title="open multiline snackbar" />
        </View>
      </StoryContainer>

      <StoryContainer title="Multiline with Button">
        <View style={{ backgroundColor: 'gainsboro', flex: 1, height: 200 }}>
          <Snackbar message={multiMessage} open={withButton} onClose={() => setWithButton(false)} type="standard" buttonText="ok" />
          <Button onPress={() => setWithButton(!withButton)} title="open multiline snackbar with button" />
        </View>
      </StoryContainer>
    </StoryWrapper>
  )
}

export const Interactive = () => {
  const [open, setOpen] = useState(false)

  return (
    <View style={{ backgroundColor: 'gainsboro', flex: 1, height: 200 }}>
      <Snackbar
        message={text('message', 'A short snackbar message')}
        buttonText={text('button text', 'ok')}
        type={select('type', snackbarTypeOptions, 'standard') as SnackbarType}
        autoHideDuration={number('auto hide duration', 5000)}
        open={open}
        onClose={() => setOpen(false)}
      />
      <Button onPress={() => setOpen(!open)} title="open snackbar" />
    </View>
  )
}
