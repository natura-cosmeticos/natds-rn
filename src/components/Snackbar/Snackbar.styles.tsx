import styled from 'styled-components/native';
import { TextProps, NativeSyntheticEvent, TargetedEvent } from 'react-native';
import { SnackbarType } from './Snackbar';
import { Theme, getSpacingSmall, getRadiusBySize } from '../../common/themeSelectors';

const getBackgroundColorByType = (theme: Theme, type: SnackbarType) => {
  switch (type) {
    case 'info':
      return theme.color.link;
    case 'warning':
      return theme.color.warning;
    case 'error':
      return theme.color.alert;
    case 'success':
      return theme.color.success;
    default:
      return theme.color.onSurface;
  }
};

export const getColorByType = (theme: Theme, type: SnackbarType) => {
  switch (type) {
    case 'info':
      return theme.color.onLink;
    case 'warning':
      return theme.color.onWarning;
    case 'error':
      return theme.color.onAlert;
    case 'success':
      return theme.color.onSuccess;
    default:
      return theme.color.surface;
  }
};


interface SnackbarButtonWrapperProps { isTwoLineAction?: boolean; }

export const SnackbarButtonWrapper = styled.View<SnackbarButtonWrapperProps>`
  margin-top: ${({ isTwoLineAction }) => (isTwoLineAction ? '0px' : '4px')};
  margin-bottom: 4px;
  margin-right: 4px;
  margin-left: 4px;
`;

interface SnackbarWrapperProps {
  theme: Theme;
  type: SnackbarType;
}

export const SnackbarWrapper = styled.View<SnackbarWrapperProps>(({ theme, type }) => ({
  alignItems: 'center',
  backgroundColor: getBackgroundColorByType(theme, type),
  borderRadius: getRadiusBySize(theme, 'medium'),
  bottom: 0,
  flexDirection: 'row',
  flexWrap: 'wrap',
  justifyContent: 'flex-end',
  left: 0,
  marginBottom: getSpacingSmall(theme),
  marginLeft: getSpacingSmall(theme),
  marginRight: getSpacingSmall(theme),
  position: 'absolute',
  right: 0,
}));

// begin - onTextLayout missing typings @0.62.2
// TODO: remove this block when upgrading @types/react-native
interface TextLayoutLine {
  ascender: number;
  capHeight: number;
  descender: number;
  height: number;
  text: string;
  width: number;
  x: number;
  xHeight: number;
  y: number;
}

interface TextLayoutEventData extends TargetedEvent {
  lines: TextLayoutLine[];
}
// end - onTextLayout missing typings @0.62.2

interface SnackbarTextProps extends TextProps {
  /**
   * Invoked on Text layout
   */
  onTextLayout?: (event: NativeSyntheticEvent<TextLayoutEventData>) => void;
  type: SnackbarType;
  isTwoLineAction?: boolean;
}

export const SnackbarText = styled.Text<SnackbarTextProps>`
  color: ${({ theme, type }) => getColorByType(theme, type)};
  flex-grow: 1;
  padding-bottom: ${({ isTwoLineAction }) => (isTwoLineAction ? 8 : 16)}px;
  padding-left: 16px;
  padding-right: 16px;
  padding-top: 16px;
  font-size: 14px;
`;
