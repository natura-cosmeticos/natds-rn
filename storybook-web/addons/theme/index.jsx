import React, { useEffect, Fragment } from 'react';
import { useAddonState, Consumer } from '@storybook/api';
import { TabButton, WithTooltip, TooltipLinkList } from '@storybook/components';
import {
  PANEL_ID, CHANGE, PARAM_KEY, getStoryBookTheme,
} from './shared';
import { light, dark } from '../../theme';

import './styles.css';

const DEFAULT_THEME = {
  name: 'natura',
  type: 'light',
};

const buildThemeProps = mode => ({ ...DEFAULT_THEME, type: mode });

function parseTheme(themes) {
  return Object.entries(themes).reduce((accum, [name, variants]) => {
    Object.entries(variants).forEach(([type]) => {
      accum.push({ name, type });
    });

    return accum;
  }, []);
}

const createBackgroundSelectorItem = ({ name, type }, onChange, selectedItem) => ({
  active: selectedItem.name === name && selectedItem.type === type,
  id: `${name} - ${type}`,
  onClick: () => onChange({ name, type }),
  title: `${name} - ${type}`,
});

const getDisplayedItems = (list, onChange, selectedItem) => {
  if (!list.length) {
    return [];
  }

  return [
    ...list.map(params => createBackgroundSelectorItem(params, onChange, selectedItem)),
  ];
};

const mapper = ({ api, state }) => {
  const story = state.storiesHash[state.storyId];
  const { themes = {} } = (story && api.getParameters(story.id, PARAM_KEY)) || {};

  return { items: parseTheme(themes) };
};

export default function Theme({ channel, api }) {
  const apiTheme = getStoryBookTheme();
  const initialTheme = apiTheme ? buildThemeProps(apiTheme) : DEFAULT_THEME;
  const [currentTheme, changeTheme] = useAddonState(PANEL_ID, initialTheme);
  const { disabled } = api.getCurrentParameter(PARAM_KEY) || {};

  const handleChange = (params) => {
    changeTheme(params);
    api.setOptions({
      theme: params.type === 'light' ? light : dark,
    });
    const event = new Event('CHANGE_THEME', { detail: params.type });

    window.dispatchEvent(event);
    channel.emit(CHANGE, params);
  };

  const renderTooltip = items => ({ onHide }) => (
    <TooltipLinkList
      links={getDisplayedItems(items, (index) => {
        handleChange(index);
        onHide();
      }, currentTheme)}
    />
  );

  const eventListener = event => handleChange(buildThemeProps(event.matches ? 'dark' : 'light'));

  useEffect(() => {
    handleChange(initialTheme);

    const matchMediaEnvent = window.matchMedia('(prefers-color-scheme: dark)');

    matchMediaEnvent.addEventListener('change', eventListener);

    return () => {
      matchMediaEnvent.removeEventListener('change', eventListener);
    };
  }, []);

  if (disabled) return null;

  return (
    <Consumer filter={mapper}>
      {({ items }) => (items.length ? (
          <Fragment>
            <WithTooltip
              placement="top"
              trigger="click"
              closeOnClick
              tooltip={renderTooltip(items)}
            >
              <TabButton
                key="Theme"
                title="Change the Theme">
                {currentTheme.name} - {currentTheme.type}
              </TabButton>
            </WithTooltip>
          </Fragment>
      ) : null)}
    </Consumer>
  );
}
