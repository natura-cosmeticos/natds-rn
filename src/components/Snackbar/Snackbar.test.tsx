import React from 'react';
import '@testing-library/jest-native/extend-expect';
import {
  act,
  cleanup,
  fireEvent,
  render,
} from '@testing-library/react-native';

import { ThemeProvider } from 'styled-components/native';

import theme from '../../common/themeSelectors/theme/mock-theme.json';
import { Snackbar, SnackbarProps } from './Snackbar';

jest.mock('react-native/Libraries/Components/Touchable/TouchableHighlight',
  () => 'TouchableHighlight');

const defaultProps = {
  message: 'A short snackbar message',
  onClose: jest.fn(),
  open: true,
  type: 'standard',
} as Omit<SnackbarProps, 'theme'>;

const ProviderWrapper = ({ children }) => (
  <ThemeProvider theme={theme}>
    {children}
  </ThemeProvider>
);

const setup = (propOverrides?: Partial<SnackbarProps>) => {
  const props = { ...defaultProps, ...propOverrides };

  const result = render(<Snackbar {...props} />, { wrapper: ProviderWrapper });

  return {
    props,
    result,
  };
};

// eslint-disable-next-line max-statements
describe('Snackbar', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    cleanup();
    jest.useRealTimers();
  });

  it('message is presented correctly', () => {
    const expectedMessage = 'A snackbar message';

    const { result } = setup({ message: expectedMessage });
    const element = result.queryByTestId('natds-snackbar-text');

    expect(element).toHaveTextContent(expectedMessage);
  });

  describe('when setting a type', () => {
    it('displays a warning type correctly', () => {
      const type = 'warning';

      const { result } = setup({ type });
      const wrapperElement = result.queryByTestId('natds-snackbar-wrapper');
      const textElement = result.queryByTestId('natds-snackbar-text');

      expect(wrapperElement).toHaveStyle({ backgroundColor: theme.color.warning });
      expect(textElement).toHaveStyle({ color: theme.color.onWarning });
    });

    it('displays a error type correctly', () => {
      const type = 'error';

      const { result } = setup({ type });
      const wrapperElement = result.queryByTestId('natds-snackbar-wrapper');
      const textElement = result.queryByTestId('natds-snackbar-text');

      expect(wrapperElement).toHaveStyle({ backgroundColor: theme.color.alert });
      expect(textElement).toHaveStyle({ color: theme.color.onAlert });
    });

    it('displays a success type correctly', () => {
      const type = 'success';

      const { result } = setup({ type });
      const wrapperElement = result.queryByTestId('natds-snackbar-wrapper');
      const textElement = result.queryByTestId('natds-snackbar-text');

      expect(wrapperElement).toHaveStyle({ backgroundColor: theme.color.success });
      expect(textElement).toHaveStyle({ color: theme.color.onSuccess });
    });

    it('displays a info type correctly', () => {
      const type = 'info';

      const { result } = setup({ type });
      const wrapperElement = result.queryByTestId('natds-snackbar-wrapper');
      const textElement = result.queryByTestId('natds-snackbar-text');

      expect(wrapperElement).toHaveStyle({ backgroundColor: theme.color.link });
      expect(textElement).toHaveStyle({ color: theme.color.onLink });
    });

    it('displays a standard type correctly', () => {
      const type = 'standard';

      const { result } = setup({ type });
      const wrapperElement = result.queryByTestId('natds-snackbar-wrapper');
      const textElement = result.queryByTestId('natds-snackbar-text');

      expect(wrapperElement).toHaveStyle({ backgroundColor: theme.color.onSurface });
      expect(textElement).toHaveStyle({ color: theme.color.surface });
    });
  });

  describe('when past a certain period of time', () => {
    it('hides snackbar automatically', () => {
      const autoHideDuration = 5000;

      const { props } = setup({ autoHideDuration });

      act(() => {
        jest.advanceTimersByTime(autoHideDuration);
      });

      expect(props.onClose).toHaveBeenCalled();
    });
  });

  describe('when setting a button text', () => {
    it('displays correctly', () => {
      const expectedButtonText = 'OK';

      const { result } = setup({ buttonText: expectedButtonText });
      const element = result.queryByTestId('natds-snackbar-button');

      expect(element).not.toBeNull();
      expect(element).toHaveTextContent(expectedButtonText);
    });

    describe('when pressing it', () => {
      it('dismisses the snackbar', () => {
        const expectedButtonText = 'OK';

        const { props, result } = setup({ buttonText: expectedButtonText });
        const element = result.getByTestId('natds-snackbar-button');

        fireEvent.press(element);
        expect(props.onClose).toHaveBeenCalled();
      });
    });
  });

  describe('when not setting a button text', () => {
    it('displays correctly', () => {
      const { result } = setup();
      const element = result.queryByTestId('natds-snackbar-button');

      expect(element).toBeNull();
    });
  });

  describe('when setting to be open', () => {
    it('displays correctly', () => {
      const { result } = setup({ open: true });
      const element = result.queryByTestId('natds-snackbar-wrapper');

      expect(element).not.toBeNull();
    });

    describe('when setting it back to be closed', () => {
      it('displays correctly', () => {
        const { props, result } = setup({ open: true });

        result.rerender(<Snackbar {...props} open={false} />);
        // eslint-disable-next-line max-nested-callbacks
        act(() => {
          jest.advanceTimersByTime(195);
        });
        const element = result.queryByTestId('natds-snackbar-wrapper');

        expect(element).toBeNull();
      });
    });
  });

  describe('when setting to be closed', () => {
    it('displays correctly', () => {
      const { result } = setup({ open: false });
      const element = result.queryByTestId('natds-snackbar-wrapper');

      expect(element).toBeNull();
    });
  });
});
