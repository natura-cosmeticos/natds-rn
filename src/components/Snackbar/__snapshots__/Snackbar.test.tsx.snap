// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snackbar should take a snapshot 1`] = `
Object {
  "props": Object {
    "message": "A short snackbar message",
    "onClose": [MockFunction],
    "open": true,
    "type": "standard",
  },
  "result": <View
    collapsable={false}
    nativeID="animatedComponent"
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#333333",
        "borderRadius": 4,
        "bottom": 0,
        "elevation": 6,
        "flexDirection": "row",
        "flexWrap": "wrap",
        "justifyContent": "flex-end",
        "left": 0,
        "marginBottom": 16,
        "marginLeft": 16,
        "marginRight": 16,
        "opacity": 0,
        "position": "absolute",
        "right": 0,
        "shadowColor": "#000000",
        "shadowOffset": Object {
          "height": 6,
          "width": 0,
        },
        "shadowOpacity": 0.14,
        "shadowRadius": 10,
      }
    }
    testID="natds-snackbar-wrapper"
    type="standard"
  >
    <Text
      ellipsizeMode="tail"
      isTwoLineAction={false}
      numberOfLines={2}
      onTextLayout={[Function]}
      style={
        Array [
          Object {
            "color": "#ffffff",
            "flexGrow": 1,
            "fontFamily": "Roboto",
            "fontSize": 14,
            "fontWeight": "400",
            "letterSpacing": 0.22,
            "lineHeight": 21,
            "paddingBottom": 16,
            "paddingLeft": 16,
            "paddingRight": 16,
            "paddingTop": 16,
          },
        ]
      }
      testID="natds-snackbar-text"
      type="standard"
    >
      A short snackbar message
    </Text>
  </View>,
}
`;
